### PR TITLE
Add deepMerge to JsonObject

### PR DESCRIPTION
**Pull request by bastewart**
Also change `Json.deepMerge` to use `toIterable` rather than `toList` internally.

### CODE REVIEW

File: modules/core/shared/src/main/scala/io/circe/JsonObject.scala [hunk 1]
```diff
@@ -166,6 +166,21 @@ sealed abstract class JsonObject extends Serializable {
    */
   final def filterKeys(pred: String => Boolean): JsonObject = filter(field => pred(field._1))
 
+  /**
+   * Perform a deep merge of this JSON object with another JSON object.
+   *
+   * Objects are merged by key, values from the argument JSON take
+   * precedence over values from this JSON. Nested objects are
+   * recursed.
```
Comment (bastewart) [inline modules/core/shared/src/main/scala/io/circe/JsonObject.scala]: Just thought I should maybe link to `Json.deepMerge` here to explain for the behaviour of values being merged?

File: modules/core/shared/src/main/scala/io/circe/JsonObject.scala [hunk 2]
```diff
@@ -166,6 +166,21 @@ sealed abstract class JsonObject extends Serializable {
...
+   */
+  def deepMerge(that: JsonObject): JsonObject =
+    toList.foldLeft(that) {
```
Comment (travisbrown) [inline modules/core/shared/src/main/scala/io/circe/JsonObject.scala]: One small thing: can you use `toIterable` here? It should behave identically but would avoid some allocations.

File: modules/core/shared/src/main/scala/io/circe/JsonObject.scala (hunk 1 above)
Comment (travisbrown) [reply modules/core/shared/src/main/scala/io/circe/JsonObject.scala]: Sounds good to me. I can never remember the exact syntax for linking methods in API docs—if you can't find it easily `[[Json]]` would definitely be fine.

Comment (bastewart) [reply modules/core/shared/src/main/scala/io/circe/JsonObject.scala]: I think what I've added is the right syntax... At least IntelliJ knows what's going on!

Comment (travisbrown) [reply modules/core/shared/src/main/scala/io/circe/JsonObject.scala]: Oh, that was easier than I thought!

File: modules/core/shared/src/main/scala/io/circe/JsonObject.scala [hunk 3]
```diff
@@ -166,6 +166,23 @@ sealed abstract class JsonObject extends Serializable {
...
   final def filterKeys(pred: String => Boolean): JsonObject = filter(field => pred(field._1))
 
+  /**
+   * Perform a deep merge of this JSON object with another JSON object.
+   *
+   * Objects are merged by key, values from the argument JSON take
+   * precedence over values from this JSON. Nested objects are
+   * recursed.
+   *
+   * See [[Json.deepMerge]] for behavior of merging values that are not objects.
```
Comment (travisbrown) [inline modules/core/shared/src/main/scala/io/circe/JsonObject.scala]: We don't need to do it in this PR, but Scaladoc actually has a `@see` tag for cases like this.

Comment (bastewart) [reply modules/core/shared/src/main/scala/io/circe/JsonObject.scala]: Ah nice, wasn't aware of that as I've never actually ended up building any scaladoc docs myself...

**Comment by travisbrown** [comment]
Thanks much @bastewart!

**Pull request merged**